README.md
# @aligames/maga-open

阿里游戏 API 网关 - Node.js SDK

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![NPM download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/@aligames/maga-open.svg?style=flat-square
[npm-url]: https://npmjs.org/package/@aligames/maga-open
[travis-image]: https://img.shields.io/travis/aliplay-team/maga-client-nodejs-open.svg?style=flat-square
[travis-url]: https://travis-ci.org/aliplay-team/maga-client-nodejs-open
[codecov-image]: https://codecov.io/gh/aliplay-team/maga-client-nodejs-open/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/gh/aliplay-team/maga-client-nodejs-open
[david-image]: https://img.shields.io/david/aliplay-team/maga-client-nodejs-open.svg?style=flat-square
[david-url]: https://david-dm.org/aliplay-team/maga-client-nodejs-open
[snyk-image]: https://snyk.io/test/npm/@aligames/maga-open/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/@aligames/maga-open
[download-image]: https://img.shields.io/npm/dm/@aligames/maga-open.svg?style=flat-square
[download-url]: https://npmjs.org/package/@aligames/maga-open

提供了 API 的请求封装、摘要签名、响应解释、消息监听等功能，使用 SDK 可以轻松完成 API 的调用，API 结果的获取。

## 安装依赖

仅支持 Node.js 6.x 以上版本。

```bash
npm i @aligames/maga-open --save
```

我们的类库严格遵循 [Semver 版本规则](http://semver.org/lang/zh-CN/)，故强烈建议开发者通过 `^` 的方式引入，即:

```json
{
  "dependencies": {
    "@aligames/maga-open": "^1.0.0"
  }
}
```

## 使用方法

### 作为消费者，发起请求

- 通过 `new Sdk.Client({ key, secret })` 进行初始化，需要提供 `key`，`secret` 配置。
- 使用 `yield client.request({ service, data, options })` 来封包，发送请求给服务端，以及解析服务端返回数据。

```js
const Sdk = require('@aligames/maga-open');
const client = new Sdk.Client({
  key: '<your_app_key>',
  secret: '<your_app_secret>',
  // host: 'http://',
  // level: 'INFO',
  // timeout: 5000, // 超时时间，默认为 5000ms
  // required: true, // 遇到错误时，直接抛错还是返回错误对象，便于组合请求，默认为 true - 抛错
  // prefix: 'MAGA',
  // logfile: 'path/to/log/file',
  // logger: my_custom_logger,
  // simplify: false // 配置为 true 的时候，会简化返回的结果，推荐配置，详见下文
});

const result = yield client.request({
  // API 名称，必须用 / 开头
  service: '/api/xxx',
  // 请求的数据体
  data: {
    uid: 'tz',
  },
});

// 返回的服务端响应
console.log(result);
```

并发组合请求：

```js
const result = yield {
  api1: client.request({ service: '/api/xxx', data: { uid: 'tz' } }),
  api2: client.request({ service: '/api/yyy' }),
};

// 返回的对象中，包含 { api1: {}, api2: {} }
console.log(result);
```

错误处理：

```js
try {
  const result = yield client.request({
    service: '/api/xxx',
    data: {
      uid: 'tz',
    },
  });
} catch (err) {
  // { message, code, data }
  console.error(err);
}
```

对特定请求配置：

```js
const config = {
  // API 名称，必须用 / 开头
  service: '/api/xxx',
  // 请求的数据体
  data: {
    uid: 'tz',
  },
};

const result = yield client.request(config, {
  timeout: 100000,
  required: false,
});

// api1 是关键请求，失败了则整个失败，而 api2 请失败时，不会报错，此时 result.api2 是 error 对象
const combo = yield {
  api1: client.request({ service: '/api/xxx', data: { uid: 'tz' } }),
  api2: client.request({ service: '/api/yyy' }, { required: false }),
};
```

对响应结果进行简化处理：

```js
// 网关返回的数据结构为: `{ id, code, message, result }`
// 其中 code 和 message 是网关层的响应，代码执行到此次，已经代表网关层响应是正常的了，故这几个字段对应用层的业务处理没啥用
// 为了方便业务应用开发者，我们在这里做下简化处理，仅把 result 节点的数据进行处理返回，即 `{ id, data, state, headers }`，其中 data 和 state 分别是 result.data 和 result.state 。

const Sdk = require('@aligames/maga-open');
const client = new Sdk.Client({
  key: '<your_app_key>',
  secret: '<your_app_secret>',

  // 配置为 true 的时候，会简化返回的结果，推荐配置
  simplify: true,
});

const result = yield client.request({
  service: '/api/xxx',
  data: {
    uid: 'tz',
  },
});

// 返回简化后的数据，包含 { id, data, state, headers }
console.log(result.data);
console.log(result.state);
```

### 作为服务提供方，提供接口服务

- 通过 `new Sdk.Server({ keystore })` 进行初始化，需要提供 `keystore` 配置。
- 使用 `server.decode({ meta, payload })` 解析客户端传递过来的请求体，解析出错会抛错，请注意 `try catch` 处理。
- 使用 `server.response({ result, id, code, msg })` 对响应数据进行封包，返回 `{ meta, payload }` 供开发者使用，前者用于配置 headers，后者是 buffer 直接返回给对端。
  - `code` 是 API 层的状态码，如 `API 找不到，返回 404`，`没有权限，keystore 校验失败，返回 403`，等等。
  - `result` 节点为业务结果，数据一般为 `{ data: {}, state: { code, message }}`
  - 注意： result.state.code 和上面提到的 state 不一样，是业务的状态码


```js
const Sdk = require('@aligames/maga-open');
const server = new Sdk.Server({
  // 允许访问的 app key 和 secret 值对
  keystore: {
    '<your_app_key>': '<your_app_secret>',
  },
  // level: 'INFO',
  // prefix: 'MAGA SERVER',
  // logfile: 'path/to/log/file',
  // logger: my_custom_logger,
});


try {
  // 获取客户端发送的请求对象，需传递 headers 和 req body buffer，进行解码，解析出错会抛错，请注意 `try catch` 处理。
  const body = server.decode({ meta: ctx.headers, payload: rawBody });

  // 进行业务逻辑处理，如查询数据库
  const result = yield ctx.service.user.find({ uid: body.uid, from: 'server' });

  // 对返回结果进行封包
  const { meta, payload } = server.response({ id: id || Date.now(), code, msg, result });

  // 返回 headers
  ctx.set(meta);

  // 返回请求体，buffer
  ctx.body = payload;

} catch (err) {
  // 需要自己处理报错后的返回
  ctx.status = err.code;
  ctx.body = err;
}
```

### 调试

我们给开发者提供了 cli 命令，用于本地发起请求。

```bash
$ maga-cli request --key=<key> --secret=<secret> <payload json string>
$ maga-cli encode --key=<key> --secret=<secret>  <payload json string>
$ maga-cli decode --key=<key> --secret=<secret>  <payload base64 string>
```

示例：

```bash
$ node ./node_modules/.bin/maga-cli --key=ngclient#2dcd --secret=wqjx0iXcRw2uEXdmjlruzw003 --host="http://localhost:7001" request '{"service":"/api/csbiz.account.findUserById?ver=1.0.0","data":{"uid":"tz"}}'

$ node bin/maga-cli.js --key=ngclient#2dcd --secret=wqjx0iXcRw2uEXdmjlruzw003 --host="http://100.84.254.233:7001" decode 'Rn+Cek0ATDXYJvkDxgiJ20+wRP4XdvFKcp4XXePyj+R83W9H+yct6LEIzrlP9cw6tohaF1a1AhcXnayIv+TfY18Kr7uJ8v9mdDBx1Efc3BUtDS3LJzW3BBhXBYeQ5C0B'
```

### 质量

我们的类库严格遵循 [Semver 版本规则](http://semver.org/lang/zh-CN/)，故强烈建议开发者通过 `^` 的方式引入，即:

```json
{
  "dependencies": {
    "@aligames/maga-open": "^1.0.0"
  }
}
```

单元测试覆盖率：

```bash
=============================== Coverage summary ===============================
Statements   : 100% ( 190/190 )
Branches     : 100% ( 80/80 )
Functions    : 100% ( 19/19 )
Lines        : 100% ( 188/188 )
================================================================================
```

### 服务端开发

参见示例： https://github.com/aliplay-team/maga-client-nodejs-open-example

推荐使用我们开源的『企业级的 Node.js Web 基础框架』 - eggjs
- 官方地址：https://eggjs.org/
- 介绍：https://zhuanlan.zhihu.com/p/25860846

## 反馈

- 联系接口同学
- 或提交 [Issue](https://github.com/aliplay-team/maga-client-nodejs-open/issues)


History.md
1.0.0 / 2017-05-08
==================

  * feat: update maga host (#1)
  * docs: add README for server (#2)
  * feat: first


test/index.test.js
'use strict';

/* eslint no-case-declarations: off */

const Sdk = require('..');
const assert = require('assert');
const mm = require('mm');
const utils = require('./test-utils');

describe('test/index.test.js', () => {
  let client;
  let server;
  let localServer;
  const key = 'ngclient#2dcd';
  const secret = 'wqjx0iXcRw2uEXdmjlruzw003';

  before(function* () {
    server = new Sdk.Server({
      keystore: {
        test: 'my-test-secret',
        'ngclient#2dcd': 'wqjx0iXcRw2uEXdmjlruzw003',
      },
      // level: 'DEBUG',
    });

    localServer = yield utils.startLocalServer((req, res) => {
      res.statusCode = 200;
      res.setHeader('x-mg-appkey', key);

      switch (req.url) {
        case '/mock/request': {
          const headers = {
            'user-agent': 'maga-client-nodejs-open/v1.0',
            'x-mg-agent': 'server',
            'x-mg-alg': 'AES-128',
            'x-mg-ts': '1493891574094',
            'x-mg-nonce': '94148431',
            'x-mg-appkey': 'ngclient#2dcd',
            'x-mg-sign': '999b46c12d4f7acc5fb6d07ea0375cb9',
            'x-mg-len': '192',
            'x-mg-code': 200,
          };

          Object.keys(headers).forEach(key => res.setHeader(key, headers[key]));
          res.end(Buffer.from('Exl51OfrqdPTTTrpGTIvOvnfXLmj0RxtnzQyzm2lGj39lv/KEpKSGiliO/q5MZo1FS8Ab8HYwYY/Wu6+XdXuq/19RwYGLW6QILEmD7N48ktgivV2KAupJEDnxS4 DcQ77XM4Ilo/4bp65hgRDoHwJSb+9qu+aEzCCWoQlxQYGpv1E8KxzQpIsS0UTZW1eZhASmb8To5+pm51Dze7N1gZSaXtbSamEATuaxiIaOYRQNnBWzWqjTRkOA67aFkAlgiwJ', 'base64'));
          break;
        }

        case '/mock/response': {
          const { meta, payload } = server.response({ result: { data: { from: 'server' }, state: { code: 200, msg: 'suc' } }, id: '123', key });
          Object.keys(meta).forEach(key => res.setHeader(key, meta[key]));
          res.end(payload);
          break;
        }

        case '/mock/response/keystore': {
          const { meta, payload } = server.response({
            result: { data: { from: 'server' },
              state: { code: 200, msg: 'suc' } },
            id: '123',
            key: req.headers['x-mg-appkey'],
          });
          Object.keys(meta).forEach(key => res.setHeader(key, meta[key]));
          res.end(payload);
          break;
        }

        case '/mock/response/empty': {
          const { meta, payload } = server.response({ result: { state: { code: 200, msg: 'suc' } }, id: '123', key });
          Object.keys(meta).forEach(key => res.setHeader(key, meta[ key ]));
          res.end(payload);
          break;
        }

        case '/mock/response/legacy': {
          const { meta, payload } = server.response({ result: { from: 'missing data and state' }, id: '123', key });
          Object.keys(meta).forEach(key => res.setHeader(key, meta[key]));
          res.end(payload);
          break;
        }

        case '/mock/response/permission': {
          try {
            server.decode({
              meta: {
                'x-mg-appkey': 'not-exist',
              },
              payload: 'a string',
            });
            res.statusCode = 500;
            res.end('should not run this');
          } catch (err) {
            const { meta, payload } = err.data;
            Object.keys(meta).forEach(key => res.setHeader(key, meta[ key ]));
            res.statusCode = err.code;
            res.end(JSON.stringify(payload));
          }
          break;
        }

        case '/mock/response/invalid': {
          res.end('abc');
          break;
        }

        case '/mock/response/invalid_request': {
          try {
            server.decode({
              meta: req.headers,
              payload: 'a string',
            });
            res.statusCode = 500;
            res.end('should not run this');
          } catch (err) {
            const { meta, payload } = err.data;
            Object.keys(meta).forEach(key => res.setHeader(key, meta[key]));
            res.statusCode = err.code;
            res.end(JSON.stringify(payload));
          }
          break;
        }

        case '/mock/response/timeout': {
          setTimeout(() => res.end('should timeout'), 1000);
          break;
        }

        case '/mock/response/500': {
          res.statusCode = 500;
          res.statusMessage = 'some error';
          res.end('abc');
          break;
        }

        default:
          res.statusCode = 500;
          break;
      }
    });

    client = new Sdk.Client({
      key,
      secret,
      host: localServer,
      simplify: true,
      // level: 'DEBUG',
    });
  });

  afterEach(mm.restore);

  it('should request', function* () {
    const result = yield client.request({ service: '/mock/request', data: {}, id: '1493891574094' });
    assert(result.id);
    assert(result.state.code === 2000000);
    assert(result.data.name === 'hello, rpc');
  });

  it('should request combo', function* () {
    const result = yield {
      api1: client.request({ service: '/mock/request', data: {}, id: '1493891574094' }),
      api2: client.request({ service: '/mock/response/empty', id: '123' }),
    };
    assert(result.api1.id);
    assert(result.api1.state.code === 2000000);
    assert(result.api1.data.name === 'hello, rpc');
    assert(result.api2.id === '123');
    assert(result.api2.state.msg === 'suc');
    assert(result.api2.state.code === 200);
  });

  it('should response', function* () {
    const result = yield client.request({ service: '/mock/response', id: '123' });
    const { id, state, data } = result;
    assert(id === '123');
    assert(state.msg === 'suc');
    assert(state.code === 200);
    assert(data.from === 'server');
  });

  it('should response with empty data', function* () {
    const result = yield client.request({ service: '/mock/response/empty', id: '123' });
    const { id, state, data } = result;
    assert(id === '123');
    assert(state.msg === 'suc');
    assert(state.code === 200);
    assert(Object.keys(data).length === 0);
  });

  it('should response with legacy support', function* () {
    const result = yield client.request({ service: '/mock/response/legacy', id: '123' });
    const { id, state, data } = result;
    assert(id === '123');
    assert(state.msg === 'warning: response missing { data, state }');
    assert(state.code === 200);
    assert(data.from === 'missing data and state');
  });

  it('should response with another client', function* () {
    const client2 = new Sdk.Client({ key: 'test', secret: 'my-test-secret', host: localServer, simplify: true });
    const result = yield client2.request({ service: '/mock/response/keystore', id: '123' });
    const { id, state, data, headers } = result;
    assert(id === '123');
    assert(state.msg === 'suc');
    assert(state.code === 200);
    assert(data.from === 'server');
    assert(headers[ 'x-mg-appkey' ] === 'test');
  });

  it('should response without simplify', function* () {
    const client2 = new Sdk.Client({ key: 'test', secret: 'my-test-secret', host: localServer });
    const response = yield client2.request({ service: '/mock/response/keystore', id: '123' });
    const { id, code, message, result, headers } = response;
    const { state, data } = result;
    assert(id === '123');
    assert(message === '业务执行成功');
    assert(code === 200);
    assert(state.msg === 'suc');
    assert(state.code === 200);
    assert(data.from === 'server');
    assert(headers['x-mg-appkey'] === 'test');
  });

  it('should decode at server', () => {
    const result = server.decode({
      meta: {
        'user-agent': 'maga-client-nodejs-open/v1.0',
        'x-mg-agent': 'server',
        'x-mg-alg': 'AES-128',
        'x-mg-ts': '1493970302764',
        'x-mg-nonce': '72208078',
        'x-mg-appkey': 'ngclient#2dcd',
        'x-mg-sign': '36746f0b6af0895da76a29ffac3070ce',
        'x-mg-len': '48',
      },
      payload: Buffer.from('v+JvwEqB+6LkeOFsCQjmvkeRkOjjWYuqhqulg0lHagRI9LbEYWwWQZgxJHy1tIzI', 'base64'),
    });
    assert(result.id === '1493970302763');
    assert(result.data);
  });

  it('should decode with invalid request id', function* () {
    try {
      server.decode({
        meta: {
          'user-agent': 'maga-client-nodejs-open/v1.0',
          'x-mg-agent': 'server',
          'x-mg-alg': 'AES-128',
          'x-mg-ts': '1493970088597',
          'x-mg-nonce': '93071486',
          'x-mg-appkey': 'ngclient#2dcd',
          'x-mg-sign': 'fc34927fe6dd35850977a4a487750e34',
          'x-mg-len': '16',
        },
        payload: Buffer.from('RVPW5VwBLHG3ckA01rKBKw==', 'base64'),
      });
      assert(false, 'should not run to this line');
    } catch (err) {
      assert(err.name === 'ProtocolDecodeError');
      assert(err.code === 400);
      assert(err.message.includes('decode error, id is required'));
    }
  });

  it('should response with invalid client', function* () {
    try {
      const client2 = new Sdk.Client({ key: 'no-exist', secret: 'no-exist', host: localServer });
      yield client2.request({ service: '/mock/response/permission' });
      assert(false, 'should not run to this line');
    } catch (err) {
      assert(err.name === 'HttpStatusError');
      assert(err.id);
      assert(err.code === 403);
      assert(err.message.includes('gateway error'));
      assert(err.message.includes('app key permission denied'));
      assert(err.data.message.includes('app key permission denied'));
    }
  });

  it('should response with invalid request data', function* () {
    try {
      yield client.request({ service: '/mock/response/invalid_request' });
      assert(false, 'should not run to this line');
    } catch (err) {
      assert(err.name === 'HttpStatusError');
      assert(err.id);
      assert(err.code === 400);
      assert(err.message.includes('gateway error'));
      assert(err.message.includes('decode error'));
      assert(err.data.message.includes('decode error'));
    }
  });

  it('should response with invalid response data', function* () {
    try {
      yield client.request({ service: '/mock/response/invalid' });
      assert(false, 'should not run to this line');
    } catch (err) {
      assert(err.name === 'ProtocolDecodeError');
      assert(err.id);
      assert(err.code === -3);
      assert(err.message.includes('decode error'));
    }
  });

  it('should support required=false when error', function* () {
    const err = yield client.request({ service: '/mock/response/invalid', options: { required: false } });
    assert(err.name === 'ProtocolDecodeError');
    assert(err.id);
    assert(err.code === -3);
    assert(err.message.includes('decode error'));
  });

  it('should response with http status error', function* () {
    try {
      yield client.request({ service: '/mock/response/500' });
      assert(false, 'should not run to this line');
    } catch (err) {
      assert(err.name === 'HttpStatusError');
      assert(err.id);
      assert(err.code === -3);
      assert(err.message.includes('gateway error'));
    }
  });

  it('should response with network error', function* () {
    try {
      yield client.request({ url: 'http://fake-tz/mock/response/invalid' });
      assert(false, 'should not run to this line');
    } catch (err) {
      assert(err.name === 'RequestError');
      assert(err.id);
      assert(err.code === -1);
      assert(err.message.includes('getaddrinfo ENOTFOUND'));
    }
  });

  it('should response with network timeout error', function* () {
    try {
      yield client.request({ service: '/mock/response/timeout', options: { timeout: 20 } });
      assert(false, 'should not run to this line');
    } catch (err) {
      assert(err.name === 'ResponseTimeoutError');
      assert(err.id);
      assert(err.code === -1);
      assert(err.message.includes('Response timeout'));
    }
  });
});



test/test-utils.js
'use strict';

const http = require('http');

let localServer;
let handler;

exports.startLocalServer = fn => {
  return new Promise((resolve, reject) => {
    handler = fn || handler || function(req, res) {
      res.statusCode = 200;
      res.end(`${req.method} ${req.url}`);
    };
    if (localServer) {
      return resolve('http://127.0.0.1:' + localServer.address().port);
    }
    localServer = http.createServer((req, res) => {
      req.resume();
      req.on('end', () => {
        handler(req, res);
      });
    });

    localServer.listen(0, err => {
      if (err) return reject(err);
      return resolve('http://127.0.0.1:' + localServer.address().port);
    });
  });
};
process.once('exit', () => localServer && localServer.close());


test/logger.test.js
'use strict';

/* eslint no-case-declarations: off */

const Sdk = require('..');
const assert = require('assert');
const path = require('path');
const rimraf = require('mz-modules/rimraf');
const sleep = require('mz-modules/sleep');
const fs = require('fs');

describe('test/index.test.js', () => {
  let client;
  const key = 'ngclient#2dcd';
  const secret = 'wqjx0iXcRw2uEXdmjlruzw003';
  const logfile = path.join(__dirname, '.tmp', 'maga.log');

  before(function* () {
    yield rimraf(path.dirname(logfile));

    client = new Sdk.Client({
      key,
      secret,
      consoleLevel: 'INFO',
      level: 'WARN',
      logfile,
    });
  });

  after(() => rimraf(path.dirname(logfile)));

  it('should log to file', function* () {
    const msg = 'test at ' + Date.now();
    client.logger.info('should not log');
    client.logger.error(msg);

    yield sleep(2000);

    const content = fs.readFileSync(logfile, 'utf-8');
    assert(content.replace(/\n/g, '') === msg);
  });
});


lib/client.js
'use strict';

const assert = require('assert');
const Protocol = require('./protocol');
const httpclient = require('./httpclient');
const Logger = require('./logger');
const SIMPLIFY_RESPONSE = Symbol('SIMPLIFY_RESPONSE');


module.exports = class MagaClient {
  /**
   * maga sdk
   * @param {Object} options - init options
   * @param {String} options.key - app key, required
   * @param {String} options.secret - app secret, required
   * @param {String} [options.host] - maga server url, don't end with `/`
   * @param {Number} [options.timeout=5000] - http timeout interval, default to 5000ms
   * @param {Boolean} [options.required=true] - whether throw error when got error or just return error object
   * @param {Boolean} [options.simplify] - whether simplify response data to developer, extract `result` property.
   * @param {String} [options.level=INFO] - logger level, default to `INFO`
   * @param {String} [options.prefix] - logger prefix
   * @param {String} [options.logfile] - logger file path
   * @param {Object} [options.logger] - logger instance, see https://github.com/eggjs/egg-logger
   * @param {Object} [options.httpclient] - urllib instance, see https://github.com/node-modules/urllib
   */
  constructor(options) {
    assert(typeof options === 'object', 'options is required');
    assert(options.key, 'options.key is required');
    assert(options.secret, 'options.secret is required');

    this.defaults = Object.assign({
      timeout: 5000,
      required: true,
      host: 'https://wx-maga.aligames.com',
      simplify: false,
    }, options);

    this.httpclient = options.httpclient || httpclient({
      keepAlive: true,
      freeSocketKeepAliveTimeout: 4000,
      timeout: 30000,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      enableDNSCache: false,
    });

    this.logger = options.logger || new Logger({
      prefix: options.prefix || 'MAGA',
      level: options.level || 'INFO',
      logfile: options.logfile,
    });

    this.protocol = new Protocol({ logger: this.logger });
  }

  /**
   * send request to maga server
   *
   * @example
   * const result = yield sdk.request({
   *   service: '/api/article.basic.list',
   *   data: { gameId: 12345 },
   *   headers: {},
   *   options: {},
   * });
   *
   * console.log(result.data);
   * console.log(result.state);
   * console.log(result.state.rt); // 响应时间
   * console.log(result.error);    // 当有错误的时候返回该节点, 或 `options.required: true` 时抛错
   *
   * @param {Object} config 请求参数 { service, data, headers, options }
   * @param {String} config.service  请求的 api 名称
   * @param {String} [config.url] 后端 URL, 可选, 默认为 host + service
   * @param {Object} [config.data] 请求参数
   * @param {Object} [config.options] 配置信息, 除了后端需要的参数外, 还增加了 timeout 和 required 等额外的控制.
   * @param {Number} [config.options.timeout] 超时时间, 默认为 5 秒
   * @param {Boolean} [config.options.required] 取值 true 时, 当发生错误将抛错; 否则返回带有 error 节点的 result 对象.
   * @return {Object} 查询到的数据, { id, data, state, error }
   */
  * request(config) {
    assert(typeof config === 'object', 'config is required');
    assert(!config.data || typeof config.data === 'object', 'config.data must be object');
    assert(!config.service || config.service.startsWith('/'), 'config.service must startsWith `/`');

    this.logger.debug('prepare request with %j', config);

    const id = config.id = String(config.id || Date.now());
    config.url = config.url || (this.defaults.host + config.service);

    const data = config.data || {};
    data.page = config.page || data.page;

    const { meta, payload } = this.encode({ id, data });
    const headers = Object.assign(config.headers || {}, meta);

    let response;
    try {
      // 发起请求
      this.logger.info('send request to %s', config.url);
      try {
        response = yield this.httpclient.request(config.url, {
          content: payload,
          headers,
          timeout: config.options && config.options.timeout || this.defaults.timeout,
          method: 'POST',
          rejectUnauthorized: false,
        });
      } catch (err) {
        // 网络错误处理: 链接异常（被拒绝/超时/中断)
        err.code = err.status;
        // err.message = 'network error, ' + err.message;
        err.data = config;
        throw err;
      }

      // 网关层错误
      const status = response.status;
      if (status < 200 || status >= 300) {
        let errData;
        try {
          errData = JSON.parse(response.data.toString());
        } catch (err) {
          errData = {
            code: -3,
            message: response.data.toString(),
          };
        }
        const msg = `gateway error, ${errData.message}, url=${config.url}, id=${config.id}`;
        const err = new Error(msg);
        Error.captureStackTrace(err);
        err.name = 'HttpStatusError';
        err.code = errData.code || /* istanbul ignore next */ status;
        err.data = errData;
        throw err;
      }

      // 对响应体进行解包处理
      try {
        // 业务数据解包，作为请求无需校验，故不传递 meta
        const protocolData = this.decode({ payload: response.data });

        this.logger.info('got response: %j', { id: protocolData.id, code: protocolData.code, message: protocolData.message, result: { toJSON: () => '[object Object] hidden at INFO' } });

        assert(String(protocolData.id) === config.id, `response id ${protocolData.id} is not equals request id ${config.id}`);

        // 对返回的响应数据，进行简化处理
        if (this.defaults.simplify) {
          // 覆盖掉 response.data
          response.data = protocolData;
          return this[SIMPLIFY_RESPONSE](response);
        }

        // 返回原来的数据, 加上 headers
        protocolData.headers = headers;
        return protocolData;

      } catch (err) {
        // 协议解包出错
        err.name = 'ProtocolDecodeError';
        err.message = 'decode error, ' + err.message;
        err.code = -3;
        err.data = response.data.toString('base64');
        throw err;
      }
    } catch (err) {
      err.id = err.id || config.id;
      // error && required => 抛错
      const required = config.options && config.options.hasOwnProperty('required') ? config.options.required : this.defaults.required;
      this.logger.error('got response err: %j, url: %s', err.message, config.url);
      if (required) {
        throw err;
      }
      // 否则返回错误对象
      return err;
    }
  }

  [SIMPLIFY_RESPONSE](response) {
    // 网关返回的数据结构为: `{ id, code, message, result }`
    // 其中 code 和 message 是网关层的响应，代码执行到此次，已经代表网关层响应是正常的了，故这几个字段对应用层的业务处理没啥用
    // 为了方便业务应用开发者，我们在这里做下简化处理，仅把 result 节点的数据进行处理返回，即 `{ id, data, state, headers }`，其中 data 和 state 分别是 result.data 和 result.state 。
    const protocolData = response.data;
    const result = {
      id: protocolData.id,
      state: {
        code: protocolData.code || /* istanbul ignore next */ status,
        msg: protocolData.message,
        rt: response.res && response.res.rt,
        size: response.res && response.res.size,
      },
      headers: response.headers,
    };

    if (protocolData.result.state) {
      result.data = protocolData.result.data || {};
      Object.assign(result.state, protocolData.result.state);
    } else {
      // 兼容没有 data 和 state 节点的情况
      result.data = protocolData.result;
      result.state.msg = 'warning: response missing { data, state }';
    }
    return result;
  }

  /**
   * encode to maga protocol
   * @param {Object} body - the data to package, { id, data }
   * @return {Object} - { meta: Object, payload: Buffer }
   * @see Protocol#encode
   */
  encode(body) {
    return this.protocol.encode({
      body,
      key: this.defaults.key,
      secret: this.defaults.secret,
    });
  }

  /**
   * decode from maga protocol
   * @param {Object} target - decode target
   * @param {Buffer} target.payload - the data to unpackage
   * @param {Object} [target.meta] - headers to validate, if not pass, will not validate
   * @return {Object} - result data
   * @see Protocol#decode
   */
  decode({ meta, payload }) {
    return this.protocol.decode({
      meta,
      payload,
      key: this.defaults.key,
      secret: this.defaults.secret,
    });
  }
};


lib/logger.js
'use strict';

const Logger = require('egg-logger').Logger;
const ConsoleTransport = require('egg-logger').ConsoleTransport;
const FileTransport = require('egg-logger').FileTransport;
const chalk = require('chalk');

class MagaLogger extends Logger {
  constructor(options) {
    /* istanbul ignore next */
    options = options || {};
    options.level = options.level || 'INFO';
    super(options);

    this.set('console', new ConsoleTransport({
      level: options.consoleLevel || options.level,
      eol: options.eol,
      formatter(meta) {
        let msg = `[${meta.level}] ${meta.message}`;

        /* istanbul ignore else */
        if (options.prefix) {
          msg = `[${options.prefix}]${msg}`;
        }

        /* istanbul ignore if */
        if (!chalk.supportsColor) {
          return msg;
        }

        /* istanbul ignore next */
        if (meta.level === 'ERROR') {
          return chalk.red(msg);
        } else if (meta.level === 'WARN') {
          return chalk.yellow(msg);
        } else if (meta.level === 'DEBUG') {
          return chalk.gray(msg);
        }

        msg = msg.replace(/(\[[\w\-_.:]+\])/g, chalk.blue('$1'));
        return msg;
      },
    }));

    if (options.logfile) {
      console.log(options.level);
      this.set('file', new FileTransport({
        file: options.logfile,
        level: options.level,
      }));
    }
  }

  get defaults() {
    return {
      level: 'INFO',
      prefix: '',
    };
  }
}

module.exports = MagaLogger;


lib/protocol.js
'use strict';

const assert = require('assert');
const md5 = require('md5');
const is = require('is-type-of');
const crypto = require('crypto');
const Logger = require('./logger');
const ALGORITHM_KEY = 'sha256';
const ALGORITHM = 'aes-128-ecb';

class MagaProtocol {
  constructor(options = {}) {
    this.logger = options.logger || new Logger({ prefix: 'MAGA' });
  }

  /**
   * encode to maga protocol
   * @param {Object} body - the data to package, { id, data }
   * @param {String} key - app key
   * @param {String} secret - app secret
   * @return {Object} - { meta: Object, payload: Buffer }
   */
  encode({ body, key, secret }) {
    assert(body, 'body is required');
    assert(body.id, 'body.id is required');

    const ts = Date.now();
    const nonce = getRandomId(8);
    const sign = md5(key + nonce + ts + JSON.stringify(body.data || {}) + secret);

    this.logger.debug('encode input: %j', body);

    const payload = this.encrypt(JSON.stringify(body), secret);
    const meta = {
      'user-agent': 'maga-client-nodejs-open/v1.0',
      'x-mg-agent': 'server',
      'x-mg-alg': 'AES-128',
      'x-mg-ts': String(ts),
      'x-mg-nonce': nonce,
      'x-mg-appkey': key,
      'x-mg-sign': sign,
      'x-mg-len': String(payload.length),
    };

    /* istanbul ignore next */
    this.logger.debug('encode output: %j, meta: %j', { toJSON: () => payload.toString('base64') }, meta);

    return { meta, payload };
  }

  /**
   * decode from maga protocol
   * @param {Buffer} payload - the data to unpackage
   * @param {Object} [meta] - headers to validate, if not pass, will not validate
   * @param {String} key - app key
   * @param {String} secret - app secret
   * @return {Object} - result data
   */
  decode({ meta, payload, key, secret }) {
    /* istanbul ignore next */
    this.logger.debug('decode input: %j', { toJSON: () => payload.toString('base64') });

    const data = JSON.parse(this.decrypt(payload, secret));

    if (meta) {
      assert(meta['x-mg-alg'] === 'AES-128', 'x-mg-alg only support AES-128');
      assert(meta['x-mg-appkey'] === key, 'x-mg-appkey is not match');
      assert(meta['x-mg-nonce'], 'x-mg-nonce is required');
      assert(meta['x-mg-ts'], 'x-mg-ts is required');
      assert(String(meta['x-mg-len']) === String(payload.length), 'x-mg-len is not match');

      const sign = md5(key + meta['x-mg-nonce'] + meta['x-mg-ts'] + JSON.stringify(data.data) + secret);
      assert(meta['x-mg-sign'] === sign, 'x-mg-sign is not match');
    }

    this.logger.debug('decode output: %j', data);

    return data;
  }

  /**
   * encrypt input data with ase128
   * @param {String|JSON|Buffer} input - data to encrypt
   * @param {String} secret - secret key
   * @return {Buffer} encrypt result buffer
   * @private
   */
  encrypt(input, secret) {
    assert(input, 'input is required');
    assert(secret, 'secret key is required');

    if (!is.buffer(input)) {
      if (typeof input === 'object') {
        input = JSON.stringify(input);
      }
      input = new Buffer(input);
    }
    // this.logger.debug('encrypt input: %j', { toJSON: () => input.toString('base64') });

    const sha = crypto.createHash(ALGORITHM_KEY);
    sha.update(secret);
    const pwd = sha.digest().slice(0, 16);

    const cipher = crypto.createCipheriv(ALGORITHM, pwd, '');
    const result = Buffer.concat([ cipher.update(input), cipher.final() ]);
    // this.logger.debug('encrypt output: %j', { toJSON: () => result.toString('base64') });
    return result;
  }

  /**
   * decrypt result data with ase128 to plain string
   * @param {String|Buffer} data - result data, support base64 string.
   * @param {String} secret - secret key
   * @return {String} decrypt result string
   * @private
   */
  decrypt(data, secret) {
    assert(data, 'data is required');
    assert(secret, 'secret key is required');

    if (is.string(data)) {
      data = Buffer.from(data, 'base64');
    }

    // this.logger.debug('decrypt input: %j', { toJSON: () => data.toString('base64') });

    const sha = crypto.createHash(ALGORITHM_KEY);
    sha.update(secret);
    const pwd = sha.digest().slice(0, 16);

    const cipher = crypto.createDecipheriv(ALGORITHM, pwd, '');
    const result = Buffer.concat([ cipher.update(data), cipher.final() ]).toString();
    // this.logger.debug('decrypt output: %s', result);
    return result;
  }
}

/**
 * gen random id
 * @param {Number} len - string length
 * @return {String}  random string with special length
   * @private
 */
function getRandomId(len) {
  let random = '';
  for (let i = 0; i < len; i++) {
    random += Math.floor(Math.random() * 10);
  }
  return random;
}

module.exports = MagaProtocol;
